tools/walletextension/api/routes.go

```diff
@@ -216,3 +237,249 @@ func submitViewingKeyRequestHandler(walletExt *walletextension.WalletExtension,
+	var messageUserID string
+	var messageAddressHex string
+	regex := regexp.MustCompile(`^Register\s(\w+)\sfor\s(\w+)$`)
+	if regex.MatchString(message) {
+		params := regex.FindStringSubmatch(message)
+		messageUserID = params[1]
+		messageAddressHex = params[2]
+	} else {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("submitted message (%s) is not in the correct format", message).Error())
+		return
+	}
+
+	// check if userID corresponds to the one in the message
+	// todo: do we need userID in query param, because we get it already from the message
+	if userID != messageUserID || len(messageUserID) != 66 {
+		userConn.HandleError(fmt.Sprintf("User in submitted message (%s) does not match user provided in the request (%s) of is wrong size", messageUserID, userID))
+		return
+	}
+
+	// Check if the signature is valid
+
+	// prefix the message like in the personal_sign method
+	prefixedMessage := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(message), message)
+	messageHash := crypto.Keccak256([]byte(prefixedMessage))
+
+	// check the signature length
+	if len(signature) != 65 {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("signature must be 64 bytes long, but %d bytes long signature received", len(signature)).Error())
+		return
+	}
+
+	// We transform the V from 27/28 to 0/1. This same change is made in Geth internals, for legacy reasons to be able
+	// to recover the address: https://github.com/ethereum/go-ethereum/blob/55599ee95d4151a2502465e0afc7c47bd1acba77/internal/ethapi/api.go#L452-L459
+	signature[64] -= 27
+
+	// check if the message contains the same address that signed the message
+	pubKey, err := crypto.SigToPub(messageHash, signature)
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error inside SigToPub: %w", err).Error())
+		return
+	}
+
+	addressFromSignature := crypto.PubkeyToAddress(*pubKey)
+	addressFromMessage := gethcommon.HexToAddress(messageAddressHex)
+
+	// verify that message was signed by the same address as in the message
+	if addressFromSignature != addressFromMessage {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("address from signature (%s) is not the same as address from message (%s)", addressFromSignature, addressFromSignature).Error())
+		return
+	}
+
+	// register the account for that viewing key
+	userIDBytes, err := hex.DecodeString(userID[2:]) // remove 0x prefix from userID
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error decoding string (%s), %w", userID[2:], err).Error())
+		return
+	}
+	err = walletExt.Storage.AddAccount(userIDBytes, addressFromMessage.Bytes(), signature)
+	if err != nil {
+		userConn.HandleError("Internal error")
+		walletExt.Logger().Error(fmt.Errorf("error while storing account (%s) for user (%s): %w", addressFromMessage.Hex(), userID, err).Error())
+		return
+	}
```

Comment:
This should be a method of the wallet extension.

Reply:
moved to WE